sympy/solvers/ode/tests/test_riccati.py

```diff
@@ -0,0 +1,950 @@
+    ====
+
+    This is the code to generate random test
+    cases and test the transformations.
+
+    for i in range(10):
+        w, b1, b2 = [rand_rational_function(x, 10, 5).cancel() for i in range(3)]
+        assert w == riccati_inverse_normal(riccati_normal(w, x, b1, b2), x, b1, b2).cancel()
+        print("    (")
+        print("        ", w, end=",\n")
+        print("        ", b1, end=",\n")
+        print("        ", b2, end=",\n")
+        print("    ),")
+    """
+    test_cases = [
+    (
+        (-50*x**4 - 12*x**3 - 30*x**2 - 30*x + 24)/(12*x**4 + 45*x**3 + 30*x**2 - 6*x),
+        (-50*x**3 + 40*x**2 - 5*x - 60)/(30*x**3 - 12*x**2 + 25*x + 80),
+        (-300*x**9 - 60*x**8 + 120*x**7 + 48*x**6 + 40*x**5 + 300*x**4 - 150*x**3 + 300*x**2 + 30*x \
+        - 36)/(45*x**9 - 60*x**8 + 30*x**7 + 40*x**6 + 60*x**5 + 60*x**4 - 90*x**3 - 12*x - 100),
+    ),
+    (
+        (-6*x**5 - 40*x**4 + 30*x**3 + 6*x**2 - 30*x - 12)/(15*x**5 + 30*x**4 + 15*x**3 - 50*x**2 + \
+        6*x - 60),
+        (18*x**3 + 16*x**2 - 12*x - 24)/(12*x**3 + 24*x**2 + 15*x - 24),
+        (-6*x**4 - 6*x**3 + 15*x**2 + 12*x)/(6*x**4 + 15*x**3 - 6*x**2 - 6*x + 2),
+    ),
+    (
+        (20*x**3 - 6*x**2 - 3)/(20*x**3 + 12*x**2 - 12*x + 12),
+        (36*x**8 + 150*x**7 + 120*x**6 - 180*x**5 - 45*x**4 - 48*x**3 + 30*x**2 - 24*x + 60)/(240*x**8 \
+        - 75*x**7 - 36*x**6 + 48*x**5 + 60*x**4 - 100*x**3 - 240*x**2 - 240*x - 12),
+        (12*x**6 - 80*x**5 + 15*x**4 + 24*x**3 + 60*x**2 - 180)/(12*x**6 - 75*x**5 + 120*x**4 - 30*x**3 \
+        - 24*x**2 + 20),
+    ),
+    (
+        (60*x**5 + 60*x**4 - 48*x**3 - 60*x**2 + 20*x + 30)/(80*x**5 - 40*x**4 - 75*x**3 - 75*x**2 - \
+        30*x + 100),
+        (12*x**5 - 2*x**4 - 2*x - 1)/(15*x**5 + x**4 - 3*x**3 + 5*x**2 - 6*x - 4),
+        (-30*x**6 - 80*x**5 + 48*x**4 - 90*x**3 - 60*x + 24)/(240*x**6 + 48*x**5 + 30*x**4 - 48*x**3 \
+        - 40*x**2 + 180*x - 45),
+    ),
+    (
+        (-180*x**10 - 120*x**9 + 36*x**8 + 120*x**7 + 60*x**6 - 120*x**4 + 30*x**3 - 90*x**2 + 60)/( \
+        100*x**9 + 15*x**7 - 80*x**6 + 48*x**5 + 240*x**4 - 150*x**3 + 60*x + 60),
+        (20*x + 4)/(3*x + 6),
+        (36*x**4 + 20*x**3 - 30*x**2 + 60*x)/(45*x**4 + 60*x**3 + 48*x**2 - 60),
+    ),
+    (
+        (4*x**3 + 4*x**2 + 16*x + 15)/(50*x**2 + 25*x + 12),
+        (-100*x**8 + 36*x**7 + 120*x**5 + 80*x**4 - 30*x**3 + 40*x**2 - 20*x - 40)/(180*x**9 + 100*x**7 \
+        - 15*x**6 + 100*x**5 + 90*x**4 + 40*x**3 + 60*x**2 + 60*x + 120),
+        (75 - 50*x)/(20*x - 6),
+    ),
+    (
+        (40*x**5 - 180*x**3 - 60*x**2 - 60*x + 300)/(180*x**6 + 150*x**5 - 100*x**4 - 48*x**3 + 45*x**2 \
+        - 20*x + 300),
+        (-150*x - 12)/(300*x**2 + 15*x - 20),
+        (-20*x**10 - 75*x**9 + 120*x**8 + 120*x**7 - 60*x**6 + 120*x**5 - 40*x**4 - 48*x**3 - 80*x**2 + \
+        60*x - 30)/(120*x**10 - 30*x**9 - 120*x**8 - 120*x**7 + 150*x**6 - 60*x**5 - 15*x**4 - 150*x**3 - \
+        150*x**2 + 120*x + 45),
+    ),
```

Comment:
It would be better if the tests were not randomly generated functions like this. Randomised testing is very useful but there are two ways it should be used:

1. The test suite can be random so every time you run the tests random inputs are generated.
2. Use randomised testing during development/debugging in order to identify problem cases.

The advantage of type 1 is that each time the test suite different examples are being tested so if the test suite runs often then it might eventually pick up on very rare cases. This can be useful but it needs careful tuning with seeds etc so that any failures can be debugged after.

I would say that what you are doing is case 2. here. In that case the basic way to write tests is like this:

1. (before writing any code!) Think up some careful examples of simple base cases and corner cases and add those to the tests.
2. Manually work through some reasonably simple examples of generic cases that cover all the possible outcomes of the algorithm/routine and add those.
3. Write the actual code
4. Run tests and measure coverage - if any lines are not covered then think carefully about test cases that should hit them and what the answer should be. Add those tests.
5. If you think that the code is correct and it seems to have full test coverage use randomised testing to identify potential bugs. Once you have identified a bug *manually* come up with a simple test case for it and add that to the tests.

Note that we don't use the randomised testing until the last step. Sometimes it could be used in step 4 if you are unable to come up with an example to get complete coverage. The same principle applies though that you are trying to identify the case so that you can use that to make a simple example.

Importantly as much as possible the tests that are written down should be hand-written. When the test suite is made up of randomly generated tests then it becomes very hard for a *human* to understand what is being tested.

In this particular test function the underlying functions being tested are very simple so they really don't need to be tested with many generic examples. I would probably test this with one fairly generic case and then any extra tests that are needed for corner cases like one of the functions being zero or anything that should raise an error.

If you look carefully at the way this function tests `ricatti_normal` you can see that although many examples are being tested not one of them tests the output of `ricatti_normal` directly.

Reply:
Okay, I will add some manually verified examples which are not complex like the ones present here.